### ISSUE TITLE
Duplicates in systemd DB causes memory leak

### ISSUE THREAD

**Issue by dsbrng25b** [fluent/fluent-bit#1864]
# Description
fluent-bit is used in Kubernetes to forward system (systemd_in) and audit logs (tail_in).
The memory usage of fluent-bit on some k8s masters is constantly growing until the fluent-bit process gets killed because the pod hits its memory resource limit (1Gi).

On one cluster all three masters are affected on another only 1 of 3 masters is affected and the other two behave normally. This is very strange since all masters run with the exact same setup. This leads me to the assumption that the problem is due to a bad state in the DB file.

After I removed the DB file and restarted fluent-bit on the master where fluent-bit was leaking memory the memory consumption remained constant (see graph).

![fluent_bit_memory_usage](https://user-images.githubusercontent.com/19696698/72078011-8f8ecc80-32f8-11ea-9f20-6cecb87e4d23.PNG)

First I assumed the problem comes from the in_tail plugin because it only appears on k8s master servers where also audit logs are forwarded. But in the DB of a bad k8s master I see a lot of duplicates in the in_systemd_cursor table. In this [ZIP](https://github.com/fluent/fluent-bit/files/4040974/sqlite_dumps.zip) you find the dumps of good and bad DBs.

On a second affected master I only removed the duplicates in the in_systemd_cursor table and restarted fluent-bit which fixed the problem.
```
DELETE FROM in_systemd_cursor WHERE rowid NOT IN ( SELECT MIN(rowid) FROM in_systemd_cursor GROUP BY cursor, updated);
```

# setup
* tested with fluent-bit versions: 1.3.1, 1.3.5
* os: Ubuntu: 18.04.1, kernel: 5.0.0-37-generic

# configuration

* `master.conf`
```
@INCLUDE main.conf
@INCLUDE systemd_input.conf
@INCLUDE kube_audit_input.conf
@INCLUDE output.conf
```
* `node.conf`
```
@INCLUDE main.conf
@INCLUDE systemd_input.conf
@INCLUDE output.conf
```

* `main.conf`
```
[SERVICE]
    Flush         1
    Log_Level     info
    Daemon        off
    Config_Watch  true
    HTTP_Server   On
    HTTP_Listen   0.0.0.0
    HTTP_Port     2020
    Parsers_File  parser.conf
```

* `kube_audit_input.conf`
```
[INPUT]
    Name    tail
    Tag     kube-audit
    Parser  kube-audit
    Path    /var/log/kubernetes/audit.log
    DB      /var/log/fluent-bit.db
    Skip_Long_Lines On
    Buffer_Max_Size 10m
    Mem_Buf_Limit 50MB
```

* `systemd_input.conf`
```
[INPUT]
    Name              systemd
    Tag               systemd
    DB                /var/log/fluent-bit.db
```

* `output.conf`
```
# do not log
[FILTER]
    Name        grep
    Match       systemd
    Exclude     MESSAGE ^Removed slice libcontainer
    Exclude     MESSAGE ^Created slice libcontainer

# splunk expects data under the key event
[FILTER]
    Name        nest
    Match       *
    Operation   nest
    Wildcard    *
    Nest_under  event

# add event metadata (index and hostname)
[FILTER]
    Name   modify
    Match  systemd
    Add    index      ${SPLUNK_INDEX_SYSTEMD}
    Add    host       ${NODE_NAME}
    Add    sourcetype systemd

[FILTER]
    Name   modify
    Match  kube-audit
    Add    index      ${SPLUNK_INDEX_KUBE_AUDIT}
    Add    host       ${NODE_NAME}
    Add    sourcetype kube_audit


[OUTPUT]
    Name            splunk
    Match           *
    Host            ${SPLUNK_HOST}
    Port            ${SPLUNK_PORT}
    Splunk_Token    ${SPLUNK_TOKEN}
    Splunk_Send_Raw On
    TLS             On
    TLS.Verify      Off
```

* `parser.conf`
```
[PARSER]
    Name kube-audit
    Format json
    Time_Key stageTimestamp
    Time_Format %Y-%m-%dT%H:%M:%S.%L%z
    Time_Keep On
```

**Comment by edsiper** [comment]
thanks @dsbrng25b for providing useful information to address the issue! (you found the duplicates problem!). Despite the root cause of corruption could not be found (non-reproducible), I've pushed a fix that makes sure no leaks will be generated and if a corrupted database is found upon start, it will do the proper sanitization.

Fixed through a1b5344 . The fix will be backported to v1.3 series.

**Comment by dsbrng25b** [comment]
@edsiper Thank you for the fix. If you would like to further investigate the root cause I have two assumptions what causes the problem:
* slow disk: fluent-bit ran stable for months and the problem started after we moved /var filesystem (where the DB is located) from tmpfs to a disk (the disk is not really slow but for sure slower then memory)
* in_tail and in_systemd use same DB: On more then 100 nodes the problem only occurred on k8s master servers where also the in_tail plugin forwards audit logs and write to the same DB.
